### PR TITLE
replace `eval`with `expr-eval` parser in computed columns

### PR DESCRIPTION
**Pull request by gka**
Changes:

* our computed columns now use `expr-eval` instead of `eval`.
* a new function `dw.utils.htmlTemplate` has been added that can be used in visualizations to evaluate tooltip templates, replacing the use of `_.template` which relies on `eval`.
* it's important to point out that we're mostly only using the token parser from `expr-eval`. most of the default functions have been renamed (we decided to uppercase all function names to avoid name clashes with lowercase variables), and some have been re-implemented to better suite our needs. see https://www.notion.so/datawrapper/My-tasks-9de6203a575f479a9ea96f09300b027d?p=031d34ec88e9485d9449fbafd072a6d9 for a complete list of supported functions

Open questions:
- [ ] at the moment this PR isn't using [silentmatt/expr-eval](https://github.com/silentmatt/expr-eval) but my patch at [gka/expr-eval](https://github.com/gka/expr-eval/tree/patch-1) because the original repo is not including source files in the npm build which makes it impossible to remove stuff from it that we don't need. do we think this is ok, for the time until my [corresponding PR](https://github.com/silentmatt/expr-eval/pull/226) gets merged?

----
In an attempt to reduce a security risk from XSS in our computed columns feature (which allows users to create columns from a JS "formula") this PR replaces the `eval()` call with [expr-eval](https://github.com/silentmatt/expr-eval).

Upsides:
* `expr-eval` doesn't rely on `eval()` so this alone makes it a lot safer. Instead it actually parses the expression.
* The syntax is probably slightly easier to use for our users. For instance `foo ^ 2` is interpreted as `Math.pow(foo, 2)` which is what users might expect from other statistical software (e.g. R). Also logical expressions use `and`, `or`, and `not` instead of `&&`, `||`, `!`, so users can write something like `IF(foo > 4 and bar < foo, "yea", "nope")`.

Downsides:
* `expr-eval` is not exactly small. it adds +14kb to our minified `dw.js` build (5kb after gzip). A smarter way would be to create two bundles and only use the one with `expr-eval` if it's needed (or to otherwise separate it from `dw.js`).
* Even though some simple expressions like `x > 5 ? col1 : col2` are identical it is fair to assume that most existing computed columns would stop working after this update. Not sure if we want to migrate them (lot of work) or somehow use `eval()` for "old" charts -- which wouldn't exactly solve the security issue.
* `expr-eval` is also basically lacking any text and date functions, but in my test I was able to [add a few ones](https://github.com/datawrapper/chart-core/blob/expr-eval/lib/dw/utils/parser.js#L133-L150) like `substr(col, 0, 10)` and `TRIM()` for text columns and `YEAR()`, `MONTH()`, `DAY()` for date columns

### CODE REVIEW

File: lib/dw/dataset/addComputedColumns.js [hunk 1]
```diff
@@ -6,33 +6,156 @@ import arrayMean from 'd3-array/src/mean';
...
+            key,
+            formula: virtualColumns[val],
+            visited: 0,
+            computed: false,
+            dependsOn: []
+        });
+    }, new Map());
+
+    // parse formulas to detect cross-column dependencies
+    _.each(virtualColumns, (formula, name) => {
```
Comment (sto3psl) [inline lib/dw/dataset/addComputedColumns.js:51]: [suggestion]  
For a non-underscore approach you could do this:
```suggestion
    Object.entries(virtualColumns).forEach(([name, formula]) => {
```

I'm not sure which one I prefer though.

File: lib/dw/dataset/addComputedColumns.js [hunk 2]
```diff
@@ -6,33 +6,156 @@ import arrayMean from 'd3-array/src/mean';
...
+        return acc.set(key, {
+            name: val,
+            origIndex: idx,
+            key,
+            formula: virtualColumns[val],
+            visited: 0,
+            computed: false,
+            dependsOn: []
+        });
+    }, new Map());
```
Comment (sto3psl) [inline lib/dw/dataset/addComputedColumns.js:48]: Cool usage of `Map` here! What made you choose `Map` over a plain object?

File: lib/dw/utils/parser.test.js [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    t.is(expr.evaluate({ date: new Date(2017, 4, 14) }), 14);
+});
+
+test('hours() function', t => {
+    const expr = parser.parse('HOURS(date)');
+    t.is(expr.evaluate({ date: new Date(2020, 1, 1) }), 0);
+    t.is(expr.evaluate({ date: new Date(2017, 4, 14, 18, 30, 5) }), 18);
+});
+
+test('sin a^b', t => {
```
Comment (sto3psl) [inline lib/dw/utils/parser.test.js:133]: Would the parser understand `a ^ b` too?

File: lib/dw/dataset/addComputedColumns.js (hunk 2 above)
Comment (gka) [reply lib/dw/dataset/addComputedColumns.js:48]: I think it was the fact that `map.set(key, val)` returns the map itself while `obj[key] = val` returns the value so I would have needed another line of code to return `obj` :)

File: lib/dw/dataset/addComputedColumns.js (hunk 1 above)
Comment (gka) [reply lib/dw/dataset/addComputedColumns.js:51]: I think in general it's a good idea to get rid of underscore dependencies. So I guess I like `Object.entries` more

File: lib/dw/utils/parser.test.js (hunk 3 above)
Comment (gka) [reply lib/dw/utils/parser.test.js:133]: sure

**Comment by davidkokkelink** [comment]
I noticed that the old implementation does _not_ include dataset changes in the computed column computation, but the new one _does_. I think the new one makes more sense, but it might lead to inconsistent behavior with some existing charts (but I don't think it should be a huge problem)

**Comment by gka** [comment]
yes that is a new feature, requested by users. I don't think it'll be a problem as this is the expected behaviour.

**Pull request merged**